Change votes to integer

phate89 opened the pull request:
As discussed in the forum I moved "Votes" from string to integer so we can actually filter it using the actual value in smart playlists. 
I left m_strVotes because I didn't know how to avoid parsing of a single cXX column from db..

MilhouseVH commented:
This needs a JSON API version bump.

Or maybe it would be easier to keep it as a string in the JSON schema, but parse out the integer value (remove commas, periods).

Montellese commented on lib/libUPnP/Platinum/Source/Devices/MediaServer/PltMediaItem.cpp:
```diff
@@ -841,7 +841,10 @@ PLT_MediaObject::FromDidl(NPT_XmlElementNode* entry)
 
-    PLT_XmlHelper::GetChildText(entry, "votes", m_XbmcInfo.votes, didl_namespace_xbmc, 256);
+    PLT_XmlHelper::GetChildText(entry, "votes", str, didl_namespace_xbmc, 256);
+    NPT_Int32 IntValue;
```
`intValue` to be in line with the `floatValue` variable from above.

Montellese commented on xbmc/interfaces/json-rpc/schema/methods.json:
```diff
@@ -1335,7 +1335,7 @@
       { "name": "mpaa", "$ref": "Optional.String" },
       { "name": "imdbnumber", "$ref": "Optional.String" },
-      { "name": "votes", "$ref": "Optional.String" },
+      { "name": "votes", "$ref": "Optional.Integer" },
```
These changes to JSON-RPC are not acceptable because they break backwards compatibility and the `votes` property isn't that important that it warrants doing that.

Montellese commented on xbmc/interfaces/legacy/InfoTagVideo.cpp:
```diff
@@ -82,9 +82,9 @@ namespace XBMCAddon
     }
 
-    String InfoTagVideo::getVotes()
+    int InfoTagVideo::getVotes()
```
Breaks backwards compatibility of the python API.

Montellese commented on xbmc/interfaces/legacy/ListItem.h:
```diff
@@ -222,7 +222,7 @@ namespace XBMCAddon
        *     - album         : string (The Joshua Tree)
        *     - artist        : list (['U2'])
-       *     - votes         : string (12345 votes)
+       *     - votes         : integer (325200)
```
According to the implementation change above you still expect a string?

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -4602,11 +4617,24 @@ void CVideoDatabase::UpdateTables(int iVersion)
+
+  if (iVersion < 97)
+  {
+    m_pDS->exec("ALTER TABLE movie ADD votes integer");
```
Where's the code that modifies the `CREAT TABLE` calls?

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -4602,11 +4617,24 @@ void CVideoDatabase::UpdateTables(int iVersion)
+    m_pDS->exec("ALTER TABLE episode ADD votes integer");
+    m_pDS->exec("ALTER TABLE tvshow ADD votes integer");
+    m_pDS->exec("ALTER TABLE musicvideo ADD votes integer");
+    std::string sql = PrepareSQL("UPDATE movie SET votes = CAST(REPLACE(c%02d,',','') as INTEGER)", VIDEODB_ID_VOTES);
```
I'm not sure if this is good enough. I have votes that look like `12,345` and others that look like `12'345` and I'm sure that with other languages/locales you'll also see other formats which are not covered by this.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -618,8 +618,8 @@ void CVideoInfoTag::ParseNative(const TiXmlElement* movie, bool prioritise)
   }
-  if (XMLUtils::GetString(movie, "votes", value))
-    SetVotes(value);
+  if (XMLUtils::GetInt(movie, "votes", after))
```
This might break existing NFOs.

Montellese commented:
Unfortunately I see a lot of API breakage due to this change which personally I don't find warranted for a field such as "votes". They could be worked around by keeping the string representation in the APIs.

MartijnKaijser commented:
@Montellese would it be an idea for adding a second "votes" field which accepts integer and keep them side by side and deprecate the string field over time?

Montellese commented:
In JSON-RPC we could change the `votes` field expected by `VideoLibrary.SetFooDetails` to support both string and integer but that doesn't work for the property returned by JSON-RPC for movies etc. There you either provide a string or an integer and providing an integer will break every client that expects a string. Same goes for Python API.

MartijnKaijser commented:
what i mean is have two fields. Keep current "votes" as string and clean it up to just have the real number and add a second field "votes_int" (or whatever name) for JSON-RPC and python that accepts integer only. That way addons and JSON client can migrate when they want to integer only.

Montellese commented:
That sounds very ugly and it will leave us with a `votes_int` (or whatever) field for the rest of the API's lifetime unless we want to break backwards compatibility at a later point and change the behaviour of `votes` to the one of `votes_int` and get rid of `votes_int` again. But nonetheless you end up breaking backwards compatibility of the API.

I would keep the string type in the API but change the value to a clean integer value and in JSON-RPC we can add a description to the field that says that this can be parsed as an integer.

MartijnKaijser commented:
yep indeed sounds ugly but was only thing that comes to mind. ok let just keep string in the API and clean it up on our side and store as integer in db. Add a big comment that they should provide a clean integer value in the string whenever possible.

Tolriq commented:
Storing as int means you need to return as String.

This trigger the problem of localization :( Thousand separator are different depending on locale.

A question I already asked before but this is the good place to ask again : 
Should the JSON API be locale agnostic or not ?

The client of a Kodi machine could use a different language / locale / timezone than the source Kodi machine so IMO API should be agnostic about that so that's the client role to format / adjust times according to it's settings and not be forced in a source setup.

Montellese commented:
@Tolriq: The stringified int would be a clean number without any thousand-separators etc so 1 million votes would be `1000000`. That's as locale agnostic as it gets.

Razzeee commented:
It should in fact be much easier for you to do different display logic, when you know you can count on getting an int.

Tolriq commented:
Yes of course :) But not breaking the API is mandatory here.

It's already a mess to handle all the JSON version let's not include incompatible things as some less advanced remotes would have troubles to handle.

topfs2 commented:
@Montellese regarding it being to small of a change to break compability, whats the plan then? Queue this up when we need to do another break anyways? Or switch to well format string and have it like that for a couple of releases, so its "safer" to break later?

phate89 commented:
Ok i fixed all the comments ( I didn't even changed the create table, dumb me) and removed the json-rpc changes and python changes. Thanks everyone.
To parse the old string I removed , . ' characters. Is it ehough? Or do  I need to create a method that only keeps numbers?
Also I'm not sure where I can parse string in json-rpc part. Is it in serialize?

Tolriq commented:
Space is also a valid thousand separator so should be removed too.

MartijnKaijser commented:
some might even insert "1652 votes" as string so i would suggest some method to only keep numbers.

phate89 commented:
Is that really a case? IF you put that you will get "1652 votes votes" everytime you show the label..

Tolriq commented:
From experience if a user can do it wrong sooner or later it will :)

MilhouseVH commented:
Ignoring all non-numeric chars would seem the safest/simplest solution.

phate89 commented:
Sure. Converting it with sql was easier but a method is safer. 
I'm not understanding the json-rpc code.Where can I apply that method for json-rpc parsing?

phate89 commented:
updated with a specific function (if you have a better name please suggest it because I had no idea how to call it)

Montellese commented on xbmc/interfaces/legacy/ListItem.cpp:
```diff
@@ -389,7 +389,9 @@ namespace XBMCAddon
             item->GetVideoInfoTag()->m_strAlbum = value;
           else if (key == "votes")
-            item->GetVideoInfoTag()->m_strVotes = value;
+          {
```
No curly braces

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -4602,11 +4617,43 @@ void CVideoDatabase::UpdateTables(int iVersion)
+    m_pDS->exec("ALTER TABLE tvshow ADD votes integer");
+    m_pDS->exec("ALTER TABLE musicvideo ADD votes integer");
+
+    std::string sql = PrepareSQL("select idMovie,cc%02d from movie", VIDEODB_ID_VOTES);
```
`c%02d` and not `cc%02d` (same below).

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -4602,11 +4617,43 @@ void CVideoDatabase::UpdateTables(int iVersion)
+    m_pDS->query(sql);
+    while (!m_pDS->eof())
+    {
+      m_pDS->exec(PrepareSQL("UPDATE movie SET votes=%i WHERE idMovie=%i", StringUtils::ReturnDigits(m_pDS->fv(1).get_asString()), m_pDS->fv(0).get_asInt()));
```
You can't use `m_pDS` to do something while iterating over a query result of `m_pDS`. Use `m_pDS2`.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -424,7 +424,7 @@ void CVideoInfoTag::Serialize(CVariant& value) const
   value["plot"] = m_strPlot;
   value["title"] = m_strTitle;
-  value["votes"] = m_strVotes;
+  value["votes"] = m_iVotes;
```
You need to convert this to a string because it's passed directly to JSON-RPC.

Montellese commented on xbmc/interfaces/json-rpc/VideoLibrary.cpp:
```diff
@@ -1134,7 +1134,7 @@ void CVideoLibrary::UpdateVideoTag(const CVariant &parameterObject, CVideoInfoTa
     SetFromDBDate(parameterObject["premiered"], details.m_premiered);
   if (ParameterNotNull(parameterObject, "votes"))
-    details.SetVotes(parameterObject["votes"].asString());
+    details.m_iVotes = parameterObject["votes"].asInteger();
```
This also needs a call to `StringUtils::ReturnDigits()`.

Montellese commented:
Please also move the changes to `lib/libUPnP` into a separate commit, create a patch of that commit and add it to `lib/libUPnP/patches` in another commit.

phate89 committed:
Change votes to int

phate89 committed:
platinum: change votes to int

phate89 commented:
@Montellese  thanks as ususal for the review. It should be ok now.

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -4602,11 +4617,43 @@ void CVideoDatabase::UpdateTables(int iVersion)
+    {
+      m_pDS2->exec(PrepareSQL("UPDATE movie SET votes=%i WHERE idMovie=%i", StringUtils::ReturnDigits(m_pDS->fv(1).get_asString()), m_pDS->fv(0).get_asInt()));
+      m_pDS->next();
+    }
```
You're missing `m_pDS->close()` in all three cases.

Montellese commented:
Only saw one more thing.

phate89 commented:
Fixed

Razzeee commented:
Just some side notice. 
While I like this to get merged. I was more hoping of votes and ratings getting put into their own tables or xml. So that we can store multiple ratings (and their vote counts) and maybe even show a avg over multiple services.

phate89 commented:
I was actually working on that too (it's all from before this pr so there's still some problem addressed with this pr like votes conversion and api breakage): https://github.com/phate89/xbmc/commit/8151c4211cbb140d5fbd9ccf552c1c32a94f4b2a
But since it wasn't finished and I wasn't sure it's a good way to handle multiple ratings I moved to this simpler step for now. 
Probably not a smart thing since some change introduced here is pointless if we move everything to a separate folder..

Razzeee commented:
Small steps are better than none or not getting there at all. I'm still +1.

Should probably also provide test builds here.

Razzeee commented:
jenkins build this please

Montellese commented:
If we want to keep the changes to the db layout minimal here we could try to re-use the existing column. We would need an update step to update the value in the columns and change the automatic result mapping to point to the new property.

phate89 committed:
add patch for 4b6935a394652fe22bab531c5654134c0daf6a8a

Montellese commented:
This can be closed in favor of #8080 right?

phate89 commented:
yes that pr include the switch to integer for votes
